numpy/core/src/multiarray/arrayobject.c

```diff
@@ -534,8 +549,72 @@ array_dealloc(PyArrayObject *self)
     npy_free_cache_dim(fa->dimensions, 2 * fa->nd);
     Py_DECREF(fa->descr);
     Py_TYPE(self)->tp_free((PyObject *)self);
+
+    /* Use new macro when available, these are identical */
+#ifdef Py_TRASHCAN_BEGIN
+    Py_TRASHCAN_END;
+#else
+    Py_TRASHCAN_SAFE_END(self);
+#endif
 }
 
+
+/*
+ * Implement traversal for cyclic garbage collection. There are two main
+ * cases. Arrays that do not own their own data, must VISIT their base.
+ * Arrays which hold the actual data own the references and have to visit
+ * all items.
+ */
+static int
+array_traverse(PyArrayObject *self, visitproc visit, void *arg)
+{
+    /* Make sure we don't traverse the array before it is fully initialized */
+    if (PyArray_DATA(self) == NULL) {
+        return 0;
+    }
```

Comment:
I think this should be an assert - if we manage to reach this, then we tracked the object too soon.